contracts/origination/OriginationControllerBase.sol

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
+import "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+import "./OriginationCalculator.sol";
+
+import "../interfaces/IOriginationController.sol";
+import "../interfaces/IOriginationHelpers.sol";
+import "../interfaces/ILoanCore.sol";
+import "../interfaces/IFeeController.sol";
+import "../interfaces/IExpressBorrow.sol";
+
+import "../libraries/OriginationLibrary.sol";
+import "../libraries/FeeLookups.sol";
+import "../libraries/Constants.sol";
+
+import "../verifiers/ArcadeItemsVerifier.sol";
+
+import {
+    OC_ZeroAddress,
+    OC_SelfApprove,
+    OC_ApprovedOwnLoan,
+    OC_InvalidSignature,
+    OC_CallerNotParticipant,
+    OC_SideMismatch,
+    OC_RolloverCurrencyMismatch,
+    OC_RolloverCollateralMismatch
+} from "../errors/Lending.sol";
+
+/**
+ * @title OriginationControllerBase
+ * @author Non-Fungible Technologies, Inc.
+ *
+ * The Origination Controller Base contract provides common functionality for all
+ * origination controllers, including signature verification and access control.
+ * It establishes access roles and integrates permission management along with
+ * signature verification functions.
+ *
+ */
+contract OriginationControllerBase is
+    IOriginationControllerBase,
+    FeeLookups,
+    EIP712,
+    OriginationCalculator,
+    ReentrancyGuard,
+    AccessControlEnumerable
+{
+    // ============================================ STATE ==============================================
+    // =================== Constants =====================
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN");
+    bytes32 public constant MIGRATION_MANAGER_ROLE = keccak256("MIGRATION_MANAGER");
+    bytes32 public constant ROLLOVER_MANAGER_ROLE = keccak256("ROLLOVER_MANAGER");
```

Comment:
Remove roles from the base contract, and put them in the relevant contracts (e.g. only `CrossCurrencyRollover` needs to know about/use the `ROLLOVER_MANAGER_ROLE`). As above, it doesn't look like `OriginationControllerBase` needs `AccessControl` at all, the child contracts do